Update Dapp Frontend Guide (#1941)
stellar/stellar-docs: docs/build/apps/dapp-frontend.mdx

If none of them suit, then it might be time to...

## Make your own template

Let’s make our own template! In this example template, we use SolidJS as the JavaScript framework, but other frameworks can be used with minor modifications. The template is using the `hello world` example smart contract, and is a part of the template initialization; bindings for the `hello world` smart contract are created.

This example template is very simple, most of the work goes into creating the `initialize.js` file, which is used to take care of creating a user account, building and deploying the smart contract, and creating the smart contract TypeScript bindings.

### 1. Initialize a SolidJS project

```bash

npx degit solidjs/templates/ts soroban-template-solid
cd soroban-template-solid
npm install
npm run dev

```

The basic SolidJS is now running on localhost port 3000.

#### Dependencies

Most of the needed dependencies are already included by the SolidJS template, we just need to add three more:

```bash

npm install dotenv glob util

> 

Reviewer: I think you might have some extra line-breaks surrounding some of the bash code blocks. make sure there are no blank lines between the three backticks on either side of the code

  );
}
```

The user is created by calling the Stellar CLI command `stellar keys generate`, and funding it with Friendbot. The user’s name is fetched from the environment variables.

You can check the new user’s public key by running this CLI command:

```bash

stellar keys public-key < my-user-name >

```

With the public key, you can look up the account on [Stellar Expert](https://stellar.expert/explorer/testnet).

#### Build contracts

We want the script to build the contract, or contracts in case there are more than one, and it’s a 2-step process. First, we clean up the target folder in case there’s a previous build, and then we call the CLI command to build the contract(s).

```javascript
// Remove all previous build files
function removeFiles(pattern) {
  glob(pattern).forEach((entry) => rmSync(entry));
}

function buildAll() {
  removeFiles(`${dirname}/target/wasm32v1-none/release/*.wasm`);
  removeFiles(`${dirname}/target/wasm32v1-none/release/*.d`);
  execSync(`stellar contract build`);
  console.log("Build complete");
}
```

The helper function `removeFiles` will delete any `wasm` or `d` files in the target directory.

#### Deploy contracts

> Now that the smart contract has been built, we can deploy the smart contract to the network, so we can invoke the smart contract functions from any client, such as our SolidJS template.

Reviewer: ```suggestion
Now that the smart contract has been built, we can deploy it to the network, so we can invoke the its functions from any client, such as our SolidJS template.
```

    await deploy(wasm);
  }
}
```

The `deploy()` function will get the contract ID from the CLI call, and add the contract name, wasm file path, and contract ID to the `smartContracts[]` array.

In this example, we only use one smart contract, but it’s not uncommon to use multiple smart contracts in a dapp, so the template supports the use of multiple contracts.

##### Create bindings

The Stellar CLI has a convenient command to create an NPM package that makes it easy to call smart contract functions from a JavaScript/TypeScript-based frontend. We call the package “bindings” because that’s what it does: it binds the contract and the frontend together.

As with the contract build functions, the binding function is also capable of handling multiple contracts, so there’s a function for creating the binding package for a contract (`bind()`) and a function that calls `bind()` for each contract (`bindAll()`).

```javascript
function bind({ alias, wasm, contractid }) {
  // Create bindings for a deployed contract
  execSync(
    `stellar contract bindings typescript --contract-id ${contractid} --output-dir ${dirname}/packages/${alias} --overwrite`,
  );
  // Build the package

>   execSync(`(cd ${dirname}/packages/${alias} && npm i && npm run build)`);

Reviewer: might wanna add `&& cd ${dirname}` or something similar to the end of this command?

}

async function bindAll() {
  // Bind all deployed contracts
  for (const contract of smartContracts) {
    await bind(contract);
  }
}
```

> The `bindAll()` function iterates the `smartContracts[]` array. The reason for not just using the array of wasms, like in the `deployAll()` function, is that we need the contract ID to generate the bindings.

Reviewer: ```suggestion
The `bindAll()` function iterates the `smartContracts[]` array. The reason for not just using the array of wasms, like in the `deployAll()` function, is that we need the contract ID to invoke the generated bindings functions on the network.
```

you can actually generate bindings from a Wasm file, but this will leave you with a client instance that requires you to specify a address

#### Import bindings

The last step is to configure the smart contract bindings client. The `importContract()` function creates a TypeScript file with a script that configures a client based on the smart contract ID, the network passphrase, and the RPC URL. The client makes it easy to make calls in the frontend code to the smart contract functions.

The file is stored with the contract name as the file name, and with the `.ts` as the extension, e.g., `hello_world.ts`.

```javascript
function importContract({ alias, wasm, contractid }) {
  const outputDir = `${dirname}/src/contracts/`;

  mkdirSync(outputDir, { recursive: true });

  const importContent =
    `import * as Client from '${alias}';\n` +

>     `export default new Client.Client({\n` +

Reviewer: ```suggestion
    `import { Client } from '${alias}';\n` +
    `export default new Client({\n` +
```

}

// ###################### Main ########################

// Calling the functions in sequence
async function main() {
  createUser();
  buildAll();
  await deployAll();
  await bindAll();
  importAll();
}

main().catch((e) => {
  console.error("Initialization failed", e);
  process.exit(1);
});
```

### 4. Modify Vite config

SolidJS is using the build tool Vite, and we need to make a minor addition to the Vite configuration file (`vite.config.ts`) for the module exports to work. Add these lines to the config:

```javascript

optimizeDeps: {
   include: ['@stellar/stellar-sdk', 'hello_world'],
 },

```

### 5. Build the frontend

The template is now ready to use the smart contract and its binding through the client. Let’s build a very simple dapp, a frontend for the Hello World smart contract, with a text input field and a send button. When a user enters a text string in the input field and clicks the send button, the contract function is invoked with the text string as the argument. The returned value, a string array, is displayed in the frontend.

Here’s an example of how the code could look the existing code in the `src/App.tsx` file with this code:

```javascript

> import type { Component } from 'solid-js';

Reviewer: ````suggestion
```javascript
import type { Component } from 'solid-js';
````

some more blank lines
